File: models/LRBinsModel.php
//Copyright (c) Meta Platforms, Inc. and affiliates.
//
//This source code is licensed under the MIT license found in the
//LICENSE file in the root directory of this source tree.

<?hh

abstract class BinaryClassificationMultistageInference{
  /**
   * Returns the combined bin key in which `$data` falls according to the
   * quantile bins found in `$quantiles`.
   */
  private function dataToCombinedBin(
    vec<vec<float>> $quantiles,
    vec<float> $data,
  ): string {
    $num_quantile_bins = C\count($quantiles);
    $num_features = C\count($quantiles[0]);

    $combined_bin = "";
    for ($j = 0; $j < $num_features; $j++) {
      $quantile_bin = 0;
      for ($i = 0; $i < $num_quantile_bins; $i++) {
        if ($data[$j] > $quantiles[$i][$j]) {
          $quantile_bin++;
        }
      }
      $combined_bin .= $quantile_bin;
    }
    return $combined_bin;
  }

  /**
   * Normalizes data to the combined bin metrics found in `$weights_info`. Then,
   * performs logistic regression inference on `$data`.
   */
  private function lrInference(
    dict<string, vec<float>> $weights_info,
    vec<float> $data,
  ): float {
    $mean = vec($weights_info["mean"]);
    $std = vec($weights_info["std"]);
    $eps = vec($weights_info["eps"]);
    $weights = vec($weights_info["weights"]);
    $len_data_row = C\count($data);
    for ($i = 0; $i < $len_data_row; $i++) {
      $data[$i] = ($data[$i] - $mean[$i]) / ($std[$i] + $eps[0]);
    }
    $sum = 0.0;
    $sum += $weights[0];
    for ($i = 0; $i < $len_data_row; $i++) {
      $sum += $data[$i] * $weights[$i + 1];
    }
    $prob = 1 / (1 + Math\exp(-$sum));
    return $prob;
  }

  /**
   * Return the features values (provided by `$id_to_feature_value`) of the
   * `$feature_ids` vector in that order. Features not found are set to zero.
   * Features are normalized according to `$data_mean`, `$data_std`, and `$eps`.
   */
  private function getData(
    dict<string, string> $id_to_feature_value,
    vec<string> $feature_ids,
    vec<float> $data_mean,
    vec<float> $data_std,
    float $eps,
  ): vec<float> {
    $num_features = C\count($feature_ids);
    $data = vec[];
    for ($i = 0; $i < $num_features; $i++) {
      if (C\contains_key($id_to_feature_value, $feature_ids[$i])) {
        $feature_val = (float)$id_to_feature_value[$feature_ids[$i]];
      } else {
        $feature_val = 0.0;
      }
      $data[] = ($feature_val - $data_mean[$i]) / ($data_std[$i] + $eps);
    }
    return $data;
  }

  /**
   * This first stage inference is called first by genProbability to attempt a
   * quick and accurate inference using the LRwBins model. The function returns
   * the probability of the logistic regression model or -1 if first-stage
   * inference cannot handle the data.
   */
  private async function genFirstStageInference(
    this::TContext $context,
  ): Awaitable<?float> {
    // load in the first-stage inference model
    $config_name = "experimental/danielsjohnson/model";
    $raw_config = Configerator::getRaw($config_name);
    $json_a = fb_json_decode($raw_config, true);

    // access the loop features
    $feature_handler = await $this->genConfigBackedFeatureHandler();
    $features_wrapper = await $feature_handler->genFeaturesAndContext(
      new LooperMemoizingWrapper($context),
    );
    list($features, $_contexts) = $features_wrapper;
    $ids = $features->getIDs();

    // map feature ids to feature values
    $id_to_feature_value = dict[];
    foreach ($ids as $id) {
      $id_to_feature_value[(string)$id] =
        PHP\strval($features->getFeature($id));
    }

    // create vector of feature values in the order used to construct the
    // combined bins (the order found in `$bin_feature_ids`)
    $eps = (float)$json_a["eps"];
    $bin_feature_ids = vec($json_a["bin_feature_ids"]);
    $bin_data_mean = vec($json_a["bin_data_mean"]);
    $bin_data_std = vec($json_a["bin_data_std"]);
    $bin_data = $this->getData(
      $id_to_feature_value,
      $bin_feature_ids,
      $bin_data_mean,
      $bin_data_std,
      $eps,
    );

    // get the combined bin key of the combined bin that `$bin_data` falls into
    $quantiles = vec($json_a["quantiles"]);
    $combined_bin = $this->dataToCombinedBin($quantiles, $bin_data);

    // check if the first-stage inference model is used for this combined bin
    $combined_bin_weights = $json_a["combined_bins"];
    if (C\contains_key($combined_bin_weights, $combined_bin)) {
      // first-stage inference is used for this prediction
      // create vector of feature values in the same order as their corresponding
      // logistic regression weights (the order found in `$inference_feature_ids`)
      $inference_feature_ids = vec($json_a["inference_feature_ids"]);
      $inference_data_mean = vec($json_a["inference_data_mean"]);
      $inference_data_std = vec($json_a["inference_data_std"]);
      $inference_data = $this->getData(
        $id_to_feature_value,
        $inference_feature_ids,
        $inference_data_mean,
        $inference_data_std,
        $eps,
      );

      // access the combined bin weights and perform logistic regression
      // inference on `$inference_data`
      $prob = $this->lrInference(
        $combined_bin_weights[$combined_bin],
        $inference_data,
      );
    } else {
      // first-stage inference is not used for this combined bin
      $prob = -1.0;
    }
    return $prob;
  }
}

File: models/LRBinsModel.py
#Copyright (c) Meta Platforms, Inc. and affiliates.
#
#This source code is licensed under the MIT license found in the
#LICENSE file in the root directory of this source tree.

import json
from typing import Any, List

import numpy as np
from fblearner.flow.projects.users.danielsjohnson.model_experiments.models.Model import (
    Model,
)
from sklearn.linear_model import LogisticRegression
from torch import tensor
from torch.distributions import Categorical
from xgboost import XGBClassifier


class LRBinsModel(Model):
    """
    Binary classification model using individual logistic regression models
    for n_models of the combined bins. When the model sees a data row whose
    bin has no model, it falls back to the fallback_model to decide.
    n_bin_features is how many features are used to put the data into bins.
    n_inference_features are how many features are used in each bin's
    logistic regression model. n_bins_per_feature is how many bins each
    feature is divided into.
    """

    def __init__(
        self,
        fallback_model: Any = None,
        n_bin_features: int = 7,
        n_inference_features: int = 20,
        n_bins_per_feature: int = 2,
        n_models: int = 50,
        feature_importances: List[float] = None,
        optimize_thresholds: bool = False,
        default_threshold: float = 0.5,
        inference_on_all_bins: bool = True,
        first_stage_threshold: float = 0.006,
        xgb_model: Any = None,
        get_bin_feature_importances: bool = False,
        inference_bins=None,
        edge_interval_bounds: str = "unbounded",
    ):
        super().__init__(
            fallback_model=fallback_model,
            feature_importances=feature_importances,
            optimize_thresholds=optimize_thresholds,
            default_threshold=default_threshold,
            inference_on_all_bins=inference_on_all_bins,
            first_stage_threshold=first_stage_threshold,
            xgb_model=xgb_model,
            get_bin_feature_importances=get_bin_feature_importances,
            edge_interval_bounds=edge_interval_bounds,
        )
        self.n_bin_features = n_bin_features
        self.n_inference_features = n_inference_features
        self.n_bins_per_feature = n_bins_per_feature
        self.n_models = n_models

    def predict_proba_one_row(
        self, X_bin_row: np.ndarray, X_inference_row: np.ndarray, full_X_row: np.ndarray
    ) -> float:
        """
        Predict probability the classes according to a single row of X.
        """
        value = self.data_to_bins_map.quantiles_to_bin.get(tuple(X_bin_row), None)
        if value is None:
            prob = self.fallback_model.predict_proba(full_X_row.reshape(1, -1))[0, 1]
        else:
            self.coverage += 1
            data_mean, data_std, eps, weights = value
            X_inference_row = (X_inference_row - np.array(data_mean)) / (
                np.array(data_std) + eps
            )
            # compute the logistic regression equation
            X_inference_row_with_bias = np.hstack((1.0, X_inference_row))
            z = np.array(weights).dot(X_inference_row_with_bias)
            prob = 1 / (1 + np.exp(-z))
        return prob

    def predict_proba(self, X_test: np.ndarray) -> np.ndarray:
        """
        Predict probabilities of the classes according to X_test.
        """
        # get important features
        bin_X_test = self.get_important_features(
            X_test, self.feature_importances, self.n_bin_features
        )
        inference_X_test = self.get_important_features(
            X_test, self.feature_importances, self.n_inference_features
        )

        # get the combined bins for the test data
        bin_X_test_quantiles = self.get_combined_bins_of_data(bin_X_test)

        # evaluate the corresponding logistic regression model based on the combined bin
        self.coverage = 0
        probs = []
        for X_bin_row, X_inference_row, full_X_row in zip(
            bin_X_test_quantiles, inference_X_test, X_test
        ):
            prob = self.predict_proba_one_row(X_bin_row, X_inference_row, full_X_row)
            probs.append(prob)
        probs = np.expand_dims(np.array(probs), 1)
        self.coverage = self.coverage / len(probs)
        return np.concatenate((np.zeros_like(probs), probs), axis=-1)

    def train_model(self, X_train: np.ndarray, y_train: np.ndarray):
        """
        Use the training data to construct the
        logistic regression with bins model.
        """
        # get important features
        bin_X_train = self.get_important_features(
            X_train, self.feature_importances, self.n_bin_features
        )
        inference_X_train = self.get_important_features(
            X_train, self.feature_importances, self.n_inference_features
        )

        # get data quantiles to determine bins using the n_bin_features
        self.set_combined_bins_of_data(bin_X_train, self.n_bins_per_feature)
        bin_X_train_quantiles = self.get_combined_bins_of_data(bin_X_train)

        # put data into combined bins
        for X_bin_row, X_inference_row, y_row in zip(
            bin_X_train_quantiles, inference_X_train, y_train
        ):
            data, labels = self.data_to_bins_map.bin_training_data.get(
                tuple(X_bin_row), ([], [])
            )
            data.append(X_inference_row)
            labels.append(y_row)
            self.data_to_bins_map.bin_training_data[tuple(X_bin_row)] = (data, labels)

        # get bin data info
        for bin in self.data_to_bins_map.bin_training_data.keys():
            data, labels = self.data_to_bins_map.bin_training_data[bin]
            self.data_to_bins_map.combined_bins_lengths[bin] = (
                np.sum(labels),
                len(labels),
            )
            if self.get_bin_feature_importances:
                data = np.array(data)
                labels = np.array(labels)
                clf = XGBClassifier()
                clf.fit(data, labels)
                local_feature_importances = clf.feature_importances_
                self.data_to_bins_map.bin_feature_importances[
                    bin
                ] = local_feature_importances

        # train a logistic regression model for each combined bin and store the weights
        for key in self.data_to_bins_map.bin_training_data.keys():
            data, labels = self.data_to_bins_map.bin_training_data[key]
            data = np.array(data)
            labels = np.array(labels)
            if len(np.unique(labels)) == 2:
                clf = LogisticRegression()
                data, data_mean, data_std, eps = self.normalize_data(data)
                clf.fit(data, labels)
                self.data_to_bins_map.quantiles_to_bin[key] = (
                    list(data_mean),
                    list(data_std),
                    eps,
                    list(np.hstack((clf.intercept_[:, None], clf.coef_))[0]),
                )
            else:
                self.data_to_bins_map.quantiles_to_bin[key] = None

    def save_model_to_json(
        self,
        filename: str,
        feature_ids: List[int],
        data_mean: float,
        data_std: float,
        eps: float,
    ):
        """
        Saves the model to a json file.
        """
        dict_to_save = {}
        dict_to_save["inference_data_mean"] = list(
            self.get_important_features(
                np.expand_dims(np.array(data_mean), axis=0),
                self.feature_importances,
                self.n_inference_features,
            )[0]
        )
        dict_to_save["inference_data_std"] = list(
            self.get_important_features(
                np.expand_dims(np.array(data_std), axis=0),
                self.feature_importances,
                self.n_inference_features,
            )[0]
        )
        dict_to_save["bin_data_mean"] = list(
            self.get_important_features(
                np.expand_dims(np.array(data_mean), axis=0),
                self.feature_importances,
                self.n_bin_features,
            )[0]
        )
        dict_to_save["bin_data_std"] = list(
            self.get_important_features(
                np.expand_dims(np.array(data_std), axis=0),
                self.feature_importances,
                self.n_bin_features,
            )[0]
        )
        dict_to_save["eps"] = eps
        dict_to_save["inference_feature_ids"] = list(
            self.get_important_features(
                np.expand_dims(np.array(feature_ids), axis=0),
                self.feature_importances,
                self.n_inference_features,
            )[0]
        )
        dict_to_save["bin_feature_ids"] = list(
            self.get_important_features(
                np.expand_dims(np.array(feature_ids), axis=0),
                self.feature_importances,
                self.n_bin_features,
            )[0]
        )
        dict_to_save["quantiles"] = np.array(self.data_to_bins_map.quantiles).tolist()
        weights_without_nones = {
            "".join(map(str, k)): v
            for k, v in self.data_to_bins_map.quantiles_to_bin.items()
            if v is not None
        }
        dict_to_save["combined_bins"] = {
            k: {"mean": mean, "std": std, "eps": [eps], "weights": weights}
            for k, (mean, std, eps, weights) in weights_without_nones.items()
        }
        with open(filename, "w") as fp:
            json.dump(dict_to_save, fp, indent=4)
